apiserver/admin.go

```diff
@@ -55,33 +86,88 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
+	}
 
-	// Use the login validation function, if one was specified.
-	if a.srv.validator != nil {
-		err := a.srv.validator(req)
-		switch err {
-		case params.UpgradeInProgressError:
-			apiRoot = restrictRoot(apiRoot, upgradeMethodsOnly)
-		case AboutToRestoreError:
-			apiRoot = restrictRoot(apiRoot, aboutToRestoreMethodsOnly)
-		case RestoreInProgressError:
-			apiRoot = restrictAll(apiRoot, restoreInProgressError)
-		case nil:
-			// in this case no need to wrap authed api so we do nothing
-		default:
-			return fail, errors.Trace(err)
+	authResult, err := a.authenticate(req)
+	if err, ok := errors.Cause(err).(*common.DischargeRequiredError); ok {
+		loginResult := params.LoginResult{
+			DischargeRequired:       err.Macaroon,
+			DischargeRequiredReason: err.Error(),
 		}
+		logger.Infof("login failed with discharge-required error: %v", err)
+		return loginResult, nil
+	}
+	if err != nil {
+		return fail, errors.Trace(err)
 	}
 
-	isUser := true
+	// Fetch the API server addresses from state.
+	hostPorts, err := a.root.state.APIHostPorts()
+	if err != nil {
+		return fail, errors.Trace(err)
+	}
+
+	model, err := a.root.state.Model()
+	if err != nil {
+		return fail, errors.Trace(err)
+	}
+
+	if authResult.userLogin {
+		switch model.MigrationMode() {
+		case state.MigrationModeImporting:
+			// The user is not able to access a model that is currently being
+			// imported until the model has been activated.
+			apiRoot = restrictAll(apiRoot, errors.New("migration in progress, model is importing"))
+		case state.MigrationModeExporting:
+			// The user is not allowed to change anything in a model that is
+			// currently being moved to another controller.
+			apiRoot = restrictRoot(apiRoot, migrationClientMethodsOnly)
+		}
+	}
+
+	loginResult := params.LoginResult{
+		Servers:       params.FromNetworkHostsPorts(hostPorts),
+		ControllerTag: model.ControllerTag().String(),
+		UserInfo:      authResult.userInfo,
+		ServerVersion: jujuversion.Current.String(),
+		PublicDNSName: a.srv.publicDNSName(),
+	}
+
+	if authResult.anonymousLogin {
+		loginResult.ModelTag = model.Tag().String()
+		loginResult.Facades = filterFacades(a.srv.facades, IsAnonymousFacade)
+		apiRoot = restrictRoot(apiRoot, anonymousFacadesOnly)
+	} else if authResult.controllerOnlyLogin {
+		loginResult.Facades = filterFacades(a.srv.facades, IsControllerFacade)
+		apiRoot = restrictRoot(apiRoot, controllerFacadesOnly)
+	} else {
+		loginResult.ModelTag = model.Tag().String()
+		loginResult.Facades = filterFacades(a.srv.facades, IsModelFacade)
+		apiRoot = restrictRoot(apiRoot, modelFacadesOnly)
+	}
+
+	a.root.rpcConn.ServeRoot(apiRoot, serverError)
+
+	return loginResult, nil
+}
+
+type authResult struct {
+	anonymousLogin      bool
+	userLogin           bool
+	controllerOnlyLogin bool
+	userInfo            *params.AuthUserInfo
+}
+
+func (a *admin) authenticate(req params.LoginRequest) (*authResult, error) {
+	result := &authResult{
+		userLogin: true,
```

Comment:
it's confusing to me to set `userLogin: true` for anonymous logins as well. can we please make them mutually exclusive?